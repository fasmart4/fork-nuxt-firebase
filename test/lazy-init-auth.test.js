const { resolve } = require('path')
const fs = require('fs-extra')
const { Nuxt, Builder } = require('nuxt')
const FirebaseModule = require('..')

jest.mock('firebase/app', () => {
  const { getProxyMock } = require('proxy-mock-js')
  const spyFn = (name, fn) => jest.fn(fn)
  const session = getProxyMock({}, 'firebaseSession', spyFn)

  /* const session = new Proxy({}, {
    get(target, name) {
      return jest.fn(() => ({}))
    }
  }) */

  return {
    apps: [session],
    messaging: {
      isSupported: jest.fn()
    }
  }
})

describe('lazy-init-auth', () => {
  let nuxt
  const buildDir = resolve(__dirname, '.nuxt-lazy-init-auth')

  beforeAll(async () => {
    const config = {
      rootDir: resolve(__dirname, '..'),
      buildDir,
      srcDir: resolve(__dirname, 'fixture'),
      modules: [
        [FirebaseModule, {
          injectModule: false,
          lazy: true,
          config: {
            // REQUIRED: Official config for firebase.initializeApp(config):
            apiKey: '<apiKey>',
            authDomain: '<authDomain>',
            databaseURL: '<databaseURL>',
            projectId: '<projectId>',
            storageBucket: '<storageBucket>',
            messagingSenderId: '<messagingSenderId>',
            appId: '<appId>',
            measurementId: '<measurementId>'
          },
          services: {
            analytics: true,
            auth: {
              initialize: {
                onAuthStateChangedAction: 'onAuthStateChanged'
              }
            },
            firestore: true,
            functions: true,
            messaging: true,
            performance: true,
            database: true,
            remoteConfig: true,
            storage: true
          }
        }]
      ]
    }
    config.dev = false

    nuxt = new Nuxt(config)
    const BundleBuilder = { build: _ => _ }
    const builder = new Builder(nuxt, BundleBuilder)
    await builder.build()
  }, 60000)

  afterAll(async () => {
    await nuxt.close()
  })

  test('init auth plugin exist', async () => {
    const content = await fs.readFile(resolve(buildDir, 'firebase/service.auth.initialize.js'), { encoding: 'utf8' })
    expect(content).toMatchSnapshot()
  })
})
