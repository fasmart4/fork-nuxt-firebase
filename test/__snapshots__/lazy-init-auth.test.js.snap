// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lazy-init-auth init auth plugin exist 1`] = `
"// runs the functions once BEFORE the root Vue.js Application is instantiated.
export default ({ $fireAuthStore }) => {
  // Can only be initiated on client side
  if (!process.client) {
    return
  }

  return $fireAuthStore.subscribe()
}
"
`;
